### DISCUSSION TITLE
Mirroring Leap Motion Hands

### DISCUSSION THREAD

**Discussion by hongkai0818** [leapmotion/UnityModules#1326] [Q&A] [answered]
In our project, we would like to place our left hand in front of the leap, and mirror it so it looks like the right hand (but controlled by the left..), so that on the screen both the left hand and right (mirrored) hand will appear. 

Is there a correct and effective way to achieve this? 

We have tried the following way but it doesn't work.
1) Duplicate the hand model and service provider. Change the X scale of the hand model to -1. It will just create a distorted hand.

**Comment by MattGrayUL** [comment, marked as answer]
Hi @hongkai0818

We don't have an out-of-the-box solution to produce what you want, but I expect you would need to modify the tracking data before it is used

We have PostProcessProviders for this purpose, [here is some documentation on them](https://docs.ultraleap.com/unity-api/The-Examples/XR/hand-post-process.html)

I have mocked up a PostProcessProvider that is a start for what you are looking for, unfortunately I have found that we may have a bug with our `Transform` extension method, but you may find this script a useful start.

Essentially, it takes an existing hand (you can choose which one) and will transform all of the hand data to be flipped on the X axis by scaling by -1

I hope this is helpful!

```
using UnityEngine;
using Leap;
using Leap.Unity;

public class MirrorLeapHand : PostProcessProvider
{
    // Choose which hand should be mirrored
    public Chirality handToMirror;

    public override void ProcessFrame(ref Frame inputFrame)
    {
        Hand sourceHand = inputFrame.GetHand(handToMirror);
        Hand mirroredHand = new Hand();

        if (sourceHand != null) // ensure we have a source hand
        {
            mirroredHand = mirroredHand.CopyFrom(sourceHand);
        }

        // invert the X scale
        LeapTransform mirrorTransform = new LeapTransform(Vector3.zero, Quaternion.identity, new Vector3(-1,1,1));

        mirroredHand = mirroredHand.Transform(mirrorTransform); // transform the hand data
        mirroredHand.IsLeft = (handToMirror == Chirality.Left) ? false : true; // set the new hands chirality
        mirroredHand.Id++; // set the new hand ID

        if (sourceHand == null)
        {
            // There is no source hand, so we have nothing to mirror
            inputFrame.Hands.Clear();
        }
        else
        {
            if (inputFrame.Hands.Count == 2) // check if there is an existing other hand
            {
                for (int i = 0; i < inputFrame.Hands.Count; i++)
                {
                    if (inputFrame.Hands[i] != sourceHand)
                    {
                        // find the hand that we should be replacing and replace it with our mirrored hand
                        inputFrame.Hands[i] = mirroredHand;
                        break;
                    }
                }
            }
            else
            {
                // there is not an existing other hand, so add a new mirror hand
                inputFrame.Hands.Add(mirroredHand);
            }
        }
    }
}
```
